Add Wireguard support

brb opened the pull request:
This PR adds a native Wireguard support. Currently, the feature is restricted to Kubernetes and ClusterPool (with single podCIDR per node and IP family) IPAM, and works in the direct routing mode only (the tunneling mode is going to be supported in the future).

The feature consists of two major components - `pkg/wireguard/agent` (which is run by cilium-agent) and `pkg/wireguard/operator` (run by a cilium-operator leader).

At the high level, we create a wireguard tunnel device (`cilium_wg0`) on each node, and set an IP from a dedicated subnet (in the code we call it as a wireguard subnet). Next, we generate a private key, and announce to remote nodes its public key via the CiliumNode object annotation. In addition, we use the same object to announce the wireguard tunnel IP, podCIDR. Finally, when the agent receives a remote CiliumNode object, it sets up the wireguard tunnel to the peer by using the nodeIP as wireguard endpoint address, the wireguard tunnel IP + podCIDR as allowed-ips, and the public key.

Currently, we encrypt the following packets:
- A local pod to a remote pod.
- A NodePort BPF request which is going to be forwarded to a remote node. 

Both types of packets are marked with `MARK_MAGIC_ENCRYPT` which is used by the following IP rule and route:
```
$ ip rule show
1:      from all fwmark 0xe00/0xf00 lookup 201
[...]
$  ip route show table 201
default dev cilium_wg0
```

Steering the packet to the `cilium_wg0` device makes Wireguard to encrypt it. Note that there is no need to specify podCIDR => nodeIP routes, as Wireguard is able to derive a destination node by consulting the allowed-ips list which the agent configures.

The operator is responsible for allocating wireguard tunnel IPs and setting them in CiliumNode objects. While, the agent is responsible for configuring the tunnel device, adding peers and direct routes.

An issue to track the follow-ups: https://github.com/cilium/cilium/issues/15462.

Co-developed with @gandro.

```release-note
Integrate Wireguard for pod2pod encryption
```

Fix #7519.

gandro committed:
vendor: Add WireGuard dependencies

This commit adds the Golang dependencies used for WireGuard support in
Cilium. It pulls in golang.zx2c4.com/wireguard/wgctrl to control the
WireGuard devices on the cilium-agent and the K8s go-client retry
package for the control-plane on the operator.

Co-authored-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

gandro committed:
wireguard: Add operator code

This adds the WireGuard operator code, which is responsible to allocate
and set the WireGuard tunnel device IP address in the CiliumNode object.

The allocation subnet will be made configurable in a command-line parameter
in a subsequent commit.

Co-authored-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

gandro committed:
wireguard: Add agent code

This commit adds the agent code used to configure the Wireguard tunnel
between nodes. The code supports adding and removing peers at run-time
and the peer information is retrieved via the CiliumNode object. The
client code which invokes this functionality is added in a subsequent
commit.

The Wireguard IP address of the local node is stored globally in the
node package similar to how other local node addresses are stored.

Co-authored-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

gandro committed:
operator: Hook up code to manage Wireguard IPs

This extends the cilium-operator to allocate Wireguard IPs for each
Cilium node from a user-defined subnet.

Co-authored-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

gandro committed:
daemon: Configure Wireguard for local node

This commit adds support in the cilium-agent for configuring the local
Wireguard device and establishing the tunnel to remote peers. This is
implemented by watching for changes to the local and remote node
objects.

Wiregurad traffic to remote nodes is redirected by the ip route based on
the MARK_MAGIC_ENCRYPT mark.

Co-authored-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

gandro committed:
helm: Add configmap option to enable Wireguard

In this initial version, Wireguard cannot be configured in a more
fine-grained manner. Follow-up work will likely add more knobs to the
Helm chart to configure additional parameters.

Co-authored-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

gandro committed:
test/k8s: Add basic Wireguard test setup

The test is checking whether KPR is wired up properly when Wireguard is
enabled, i.e. whether a NodePort BPF requests when a backend is on a
remote node are forwarded over the wireguard tunnel.

The log-gatherer image vsn bump is to be able to run tcpdump from it.

Co-authored-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

tklauser reviewed (commented):
Nice 🚀 

Haven't looked at the full PR in detail, but stumbled on this bit in the description:

> The feature consists of two major componens - `pkg/wireguard/agent.go` (which is run by cilium-agent) and `pkg/wireguard/operator.go` (run by a cilium-operator leader).

Ideally (and if possible) I'd suggest to split the agent and operator components into separate packages `pkg/wireguard/agent` and `pkg/wireguard/operator`, with only the common parts (if any) residing in `pkg/wireguard`. This will be beneficial in terms of binary size since this avoids unnecessrily pulling in transitive dependencies (e.g. as far as I can see only `pkg/wireguard/agent.go` pulls in `golang.zx2c4.com/wireguard/wgctrl` while `pkg/wireguard/operator.go` doesn't).

brb committed:
datapath: Set mark to encrypt traffic with Wireguard

This commit sets MARK_MAGIC_ENCRYPT for the following packets:
- A local endpoint (pod) to a remote endpoint.
- A to-be forwarded NodePort BPF request to a remote service backend.

The marked packet is a subject to the ip rule inserted by the
Wireguard agent:

    $ ip rule show
        1:      from all fwmark 0xe00/0xf00 lookup 201
        ...
    $  ip route show table 201
    default dev cilium_wg0

The route makes the traffic to be encrypted with Wireguard (it's enough
to redirect the packet to the tunnel for it; Wireguard will decide to
which node to route the packet afterwards by consulting the
allowed-ips).

Co-authored-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

christarazi commented on pkg/wireguard/operator/operator.go:
```diff
@@ -0,0 +1,360 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operator
```
Would be good to add the description from the PR about this package as a godoc here

christarazi commented on pkg/wireguard/operator/operator.go:
```diff
@@ -0,0 +1,360 @@
+	ipv6AllocAfterRestore map[string]struct{} // by nodename
+}
+
+func NewOperator(subnetV4, subnetV6 *net.IPNet, ciliumNodeUpdater CiliumNodeUpdater) (*Operator, error) {
```
godocs for exported methods

christarazi commented on pkg/wireguard/operator/operator.go:
```diff
@@ -0,0 +1,360 @@
+		ipv6AllocAfterRestore: ipv6AllocAfterRestore,
+	}
+
+	return m, nil
```
Nit: this can be inlined

christarazi commented on pkg/wireguard/operator/operator.go:
```diff
@@ -0,0 +1,360 @@
+		ipAlloc = o.ipv6Alloc
+		ipByNode = o.ipv6ByNode
+	default:
+		log.Warning("unsupported family")
```
I assume the the user can manually modify the CiliumNode resource and make a typo here, so I think this warning should provide much more context.

gandro replied:
The family here is actually only ever set by the programmer. We decided to address this by introducing a panic here, since passing the wrong familiy is always a programmer error (and not something the user can address).

christarazi commented on pkg/wireguard/operator/operator.go:
```diff
@@ -0,0 +1,360 @@
+}
+
+func (o *Operator) setCiliumNodeIP(nodeName string, ip net.IP) error {
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```
TIL, this is pretty cool. We should start using this in our K8s code

christarazi commented on pkg/wireguard/operator/operator.go:
```diff
@@ -0,0 +1,360 @@
+
+	log.WithFields(logrus.Fields{
+		"nodeName": nodeName,
+		"ip":       ip,
```
Nit: we have `logfields` for these

christarazi commented on pkg/datapath/linux/linux_defaults/linux_defaults.go:
```diff
@@ -23,6 +23,10 @@ const (
 	// RouteTableIPSec is the default table ID to use for IPSec routing rules
 	RouteTableIPSec = 200
 
+	// RouteTableWireguard is the default table ID to use for Wireguard routin
```
typo: s/routin/routing

christarazi commented on pkg/wireguard/agent/agent.go:
```diff
@@ -0,0 +1,352 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package agent
```
Same comment here for package godoc from PR description

christarazi commented on pkg/wireguard/agent/agent.go:
```diff
@@ -0,0 +1,352 @@
+		ListenPort:   &a.listenPort,
+		ReplacePeers: false,
+	}
+	if err := a.wgClient.ConfigureDevice(IfaceName, *cfg); err != nil {
```
No need for this to be a pointer, since it's dereferenced right away anyway, correct?

christarazi commented on pkg/wireguard/agent/agent.go:
```diff
@@ -0,0 +1,352 @@
+	// Create the rule to steer the marked traffic (from a local endpoint to a
+	// remote endpoint) via the Wireguard tunnel
+	rule := route.Rule{
+		Priority: 1,
```
We should define this where we define other routing table constants. Makes it easier to track and figure out which values / constants we rely on

christarazi commented on pkg/wireguard/agent/agent.go:
```diff
@@ -0,0 +1,352 @@
+	ip, net, _ := net.ParseCIDR("0.0.0.0/0")
+	net.IP = ip
+	rt := route.Route{
+		Prefix: *net,
```
I think you can use `net.IPv4zero` for this instead of manually creating a CIDR

christarazi commented on pkg/wireguard/agent/agent.go:
```diff
@@ -0,0 +1,352 @@
+		"wgIPv4":    wgIPv4,
+		"nodeIPv6":  nodeIPv6,
+		"podCIDRv6": podCIDRv6,
+		"wgIPv6":    wgIPv6,
```
`logfields`

christarazi commented on pkg/wireguard/agent/agent.go:
```diff
@@ -0,0 +1,352 @@
+	return nil
+}
+
+func (a *Agent) UpdatePeer(nodeName, pubKeyHex string,
+	wgIPv4, nodeIPv4 net.IP, podCIDRv4 *net.IPNet,
+	wgIPv6, nodeIPv6 net.IP, podCIDRv6 *net.IPNet) error {
```
Should we define a `Peer` type? Would that make the code more flexible in the future or do you feel it's not necessary?

gandro replied:
Martynas and I discussed this, and we prefer to leave it as is for now. We will likely extend this method in a follow-up to support managed Kubernetes. Once we do that, we'll have a better understanding of how the `Peer` struct needs to look like.

christarazi commented on pkg/wireguard/agent/agent.go:
```diff
@@ -0,0 +1,352 @@
+	return nil
+}
+
+func loadOrGeneratePrivKey(filePath string) (key wgtypes.Key, err error) {
```
Should we abstract this to use `os.Reader` instead of forcing the use of a file? It will ease unit tests

gandro replied:
Not sure how we can implement the body of this function with an `os.Reader`. It needs to check if the file exists or not. It should be unit-testable via  `t.TempDir`.

christarazi commented on operator/cilium_node.go:
```diff
@@ -57,6 +60,11 @@ func startSynchronizingCiliumNodes(nodeManager allocator.NodeEventHandler) {
 				if node := k8s.ObjToCiliumNode(obj); node != nil {
 					// node is deep copied before it is stored in pkg/aws/eni
 					nodeManager.Create(node)
+					if wgOperator != nil {
```
Once https://github.com/cilium/cilium/pull/15295 is merged, we can convert this logic to be inside its own handler. It's OK to do this for now

christarazi commented on pkg/option/config.go:
```diff
@@ -1603,6 +1612,10 @@ type DaemonConfig struct {
 
+	EnableWireguard   bool
+	WireguardSubnetV4 *net.IPNet
+	WireguardSubnetV6 *net.IPNet
```
These should have godocs

christarazi replied:
Also, these should also be put under the `Networking` section for organizing the flags, right?

christarazi commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -339,6 +340,10 @@ func initKubeProxyReplacementOptions() (strict bool) {
 // handleNativeDevices tries to detect bpf_host devices (if needed).
 func handleNativeDevices(strict bool) {
+	if option.Config.EnableWireguard {
+		option.Config.DirectRoutingDevice = wireguard.IfaceName
```
Maybe it's worth extracting common defines into a `types` package within `pkg/wireguard`, as to avoid importing the entire `agent` package just for a define

christarazi commented on pkg/datapath/linux/datapath.go:
```diff
@@ -38,19 +39,21 @@ type linuxDatapath struct {
 
 // NewDatapath creates a new Linux datapath
-func NewDatapath(cfg DatapathConfiguration, ruleManager datapath.IptablesManager) datapath.Datapath {
+func NewDatapath(cfg DatapathConfiguration, ruleManager datapath.IptablesManager, wgAgent *wireguard.Agent) datapath.Datapath {
```
Shouldn't the datapath define a local interface to abstract away the real `wireguard.Agent`? This should reduce the imports of the package where it's unnecessary

gandro replied:
Good feedback!

christarazi reviewed (commented):
Nice! First round feedback below

kkourt commented on pkg/wireguard/operator/operator.go:
```diff
@@ -0,0 +1,369 @@
+	Get(node string) (*v2.CiliumNode, error)
+}
+
+type Operator struct {
```
Thanks for the well-documented code! I enjoyed reading it!

How about adding something like the following to provide a basic understanding of restoring?
```
// Operator is initialized with NewOperator()
//
// The operator starts in restoring mode, when it will not attempt to allocate
// any IPs and, instead, place nodes in ipv{4,6}AllocAfterRestore so that their
// IPs can be allocated on RestoreFinished()
```

kkourt commented on pkg/wireguard/operator/operator.go:
```diff
@@ -0,0 +1,369 @@
+	log.WithFields(logrus.Fields{
+		logfields.NodeName: nodeName,
+		logfields.IPAddr:   ip,
+	}).Info("Set wireguard IP")
```
how about adding the error here as well?

Or maybe something like this? (totally untested)
```
	l := log.WithFields(logrus.Fields{
		logfields.NodeName: nodeName,
		logfields.IPAddr:   ip,
	})

	if err != nil {
		l.Info("Set wireguard IP")
	} else {
		l.WithError(err).Warn("Set wireguard IP failed")
	}

	return err
```

kkourt commented on pkg/wireguard/operator/operator.go:
```diff
@@ -0,0 +1,369 @@
+	defer o.Unlock()
+
+	for nodeName := range o.ipv4AllocAfterRestore {
+		ip, err := o.ipv4Alloc.AllocateNext()
```
Seems to me that the entries of `ipv4AllocAfterRestore` are never deleted. This is probably fine since, AFAICT, we can never go from the non-restoring state to the restoring state, but maybe is worth removing the entries anyway?

gandro replied:
Good catch! We'll clear it to allow for GC.

rolinh commented on pkg/node/address.go:
```diff
@@ -566,6 +570,30 @@ func SetK8sNodeIP(ip net.IP) {
 	k8sNodeIP = ip
 }
 
+func SetWireguardIPv4(ip net.IP) {
+	wireguardIPv4 = ip
+}
```
Shouldn't we ensure that `ip` is actually v4 and not v6?

gandro replied:
Unless you feel strongly about this, I'll leave this as is, especially since all the other setters in this package are also very bare bones, c.f. `SetIPv4`, `SetIPv4Loopback`, `SetIPv4AllocRange`

Panicking is probably not the right choice here, and returning an error would also not be useful, as the only invocation site already necessarily has to check for IPv4.

rolinh commented on pkg/node/address.go:
```diff
@@ -566,6 +570,30 @@ func SetK8sNodeIP(ip net.IP) {
+	return wireguardIPv4
+}
+
+func SetWireguardIPv6(ip net.IP) {
+	wireguardIPv6 = ip
+}
```
Same comment here but for v6.

kkourt commented on pkg/wireguard/agent/agent.go:
```diff
@@ -0,0 +1,356 @@
+		return nil, err
+	}
+
+	node.SetWireguardPubKey(key.PublicKey().String())
```
Would it be better to do this after wgctrl.New() below in case the latter fails?

kkourt commented on pkg/wireguard/agent/agent.go:
```diff
@@ -0,0 +1,356 @@
+
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "wireguard")
+
+type Agent struct {
```
How about adding a few lines as a comment about how is this supposed to work?

E.g., 

> Agent is initialized with .Init() where the wireguard link will be created and the proper routes set. During Init, existing peer keys are placed into `restoredPubKeys`. Once `RestoreFinished` is called obsolete keys and peers are removed. `UpdatePeer` updates the public key of a peer.

(The above might not be fully correct -- it was my understanding from reading the code)

gandro replied:
Good and correct summary! Thanks, I've adapted a version of it.

rolinh reviewed (approved):
Just one question, lgtm otherwise.

kkourt commented on pkg/datapath/linux/node.go:
```diff
@@ -910,9 +913,28 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
+		if newNode.IPv6AllocCIDR != nil {
+			podCIDRv6 = newNode.IPv6AllocCIDR.IPNet
+		}
+		if err := n.wgAgent.UpdatePeer(newNode.Name, newNode.WireguardPubKey,
```
Could it be the case that `Config.EnableWireguard` is true, but the agent failed to initialize and this will cause the agent to crash? Should we check for `n.Agent == nil` here?

brb replied:
We will fatal if the WG agent failed to initialize:
```
wgAgent, err = wireguard.NewAgent(privateKeyPath,
			option.Config.WireguardSubnetV4, option.Config.WireguardSubnetV6)
		if err != nil {
			log.WithError(err).Fatal("Failed to initialize wireguard")
		}
```

kkourt commented on pkg/datapath/linux/node.go:
```diff
@@ -998,6 +1020,12 @@ func (n *linuxNodeHandler) nodeDelete(oldNode *nodeTypes.Node) error {
 	}
 
+	if option.Config.EnableWireguard {
+		if err := n.wgAgent.DeletePeer(oldNode.Name); err != nil {
```
ditto, as previous comment.

brb replied:
Same as above :-)

kkourt reviewed (commented):
Thank you for the nice PR!

I have some minor comments, and I'd like to have another look going backwards ~tomorrow~ later today.

tklauser commented on pkg/datapath/linux/linux_defaults/linux_defaults.go:
```diff
@@ -52,6 +56,9 @@ const (
 	// IPSecProtocolID IP protocol ID for IPSec defined in RFC4303
 	RouteProtocolIPSec = 50
 
+	// RulePriorityWireguard is the priority of the rule used for routing packets to Wireguard device for encryption
```
Nit: line-wrap this comment at the same width as the other comments in this `const` group.

borkmann replied:
Will this work in combination with BPF HR?

brb replied:
> Will this work in combination with BPF HR?

@borkmann Hmm, I couldn't make it work with BPF HR. For testing, I had to disable the auto setting of direct routing device to`cilium_wg0` (due to https://github.com/cilium/cilium/issues/15075), and also disable the `ENABLE_SKIP_FIB`. For some reasons, the fib_lookup() returns `oif` = non-wireguard device regardless the ip rule, which bypasses the encryption.

brb replied:
Discussed offline that previously the `fib_lookup()` with `BPF_FIB_LOOKUP_DIRECT` ignored ip rules. Unfortunately, removing the flag still seems to skip the rule. Maybe it ignores the skb mark? In the worst we could fix this by doing a redirect to the Wireguard tunnel device (added to the follow-up issue).

tklauser commented on pkg/node/types/node.go:
```diff
@@ -303,6 +307,19 @@ func (n *Node) GetCiliumInternalIP(ipv6 bool) net.IP {
+		if addr.Type == addrType {
+			return addr.IP
+		}
+	}
```
Move the address type check before the IP check, as `net.IP.To4()` might allocate:

```suggestion
	for _, addr := range n.IPAddresses {
                if addr.Type != addrType {
                        continue
                }
                if is4 := addr.IP.To4() != nil; (!ipv6 && is4) || (ipv6 && !is4) {
                        return addr.IP
                }
	}
```

FWIW, I've verified the code above using the following unit test which passes both with and without the change:

<details>
  <summary>Diff for <code>(*Node).GetIPByType</code> unit test</summary>

```diff
diff --git a/pkg/node/types/node_test.go b/pkg/node/types/node_test.go
index 6a16d5c3808c..c11e8bfd51a2 100644
--- a/pkg/node/types/node_test.go
+++ b/pkg/node/types/node_test.go
@@ -78,6 +78,60 @@ func (s *NodeSuite) TestGetNodeIP(c *C) {
 
 }
 
+func (s *NodeSuite) TestGetIPByType(c *C) {
+       n := Node{
+               Name: "node-1",
+               IPAddresses: []Address{
+                       {IP: net.ParseIP("192.0.2.3"), Type: addressing.NodeExternalIP},
+               },
+       }
+
+       ip := n.GetIPByType(addressing.NodeInternalIP, false)
+       c.Assert(ip, IsNil)
+       ip = n.GetIPByType(addressing.NodeInternalIP, true)
+       c.Assert(ip, IsNil)
+
+       ip = n.GetIPByType(addressing.NodeExternalIP, false)
+       c.Assert(ip.Equal(net.ParseIP("192.0.2.3")), Equals, true)
+       ip = n.GetIPByType(addressing.NodeExternalIP, true)
+       c.Assert(ip, IsNil)
+
+       n = Node{
+               Name: "node-2",
+               IPAddresses: []Address{
+                       {IP: net.ParseIP("f00b::1"), Type: addressing.NodeCiliumInternalIP},
+               },
+       }
+
+       ip = n.GetIPByType(addressing.NodeExternalIP, false)
+       c.Assert(ip, IsNil)
+       ip = n.GetIPByType(addressing.NodeExternalIP, true)
+       c.Assert(ip, IsNil)
+
+       ip = n.GetIPByType(addressing.NodeCiliumInternalIP, false)
+       c.Assert(ip, IsNil)
+       ip = n.GetIPByType(addressing.NodeCiliumInternalIP, true)
+       c.Assert(ip.Equal(net.ParseIP("f00b::1")), Equals, true)
+
+       n = Node{
+               Name: "node-3",
+               IPAddresses: []Address{
+                       {IP: net.ParseIP("192.42.0.3"), Type: addressing.NodeExternalIP},
+                       {IP: net.ParseIP("f00d::1"), Type: addressing.NodeExternalIP},
+               },
+       }
+
+       ip = n.GetIPByType(addressing.NodeInternalIP, false)
+       c.Assert(ip, IsNil)
+       ip = n.GetIPByType(addressing.NodeInternalIP, true)
+       c.Assert(ip, IsNil)
+
+       ip = n.GetIPByType(addressing.NodeExternalIP, false)
+       c.Assert(ip.Equal(net.ParseIP("192.42.0.3")), Equals, true)
+       ip = n.GetIPByType(addressing.NodeExternalIP, true)
+       c.Assert(ip.Equal(net.ParseIP("f00d::1")), Equals, true)
+}
+
```

</details>

Feel free to pick this up into the respective commit.

gandro replied:
Very nice suggestion, thanks a lot! Will apply in the next push!

tklauser commented on pkg/wireguard/agent/agent.go:
```diff
@@ -0,0 +1,356 @@
+		return err
+	}
+
+	if err := sysctl.Write(fmt.Sprintf("net.ipv4.conf.%s.rp_filter", types.IfaceName), "2"); err != nil {
```
Is this also needed if `option.Config.EnableIPv4 == false`?

tklauser commented on pkg/wireguard/agent/agent.go:
```diff
@@ -0,0 +1,356 @@
+		return nil
+	}
+
+	type param struct {
```
Nit: maybe give this type a more descriptive name, e.g. `ipNetFamily`?

tklauser commented on pkg/datapath/linux/config/config.go:
```diff
@@ -472,7 +476,10 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 	}
 
-	if option.Config.EnableIdentityMark {
+	if option.Config.EnableIdentityMark && !option.Config.EnableWireguard {
```
Setting both to true seems to blindly disable the identity mark. Should we instead emit a warning or info message that identity marks are disabled?

Also it looks like the default for `EnableIdentityMark` is to be set `true`, so maybe just set it to `false` in case wireguard is enabled?

brb replied:
I got rid of this check. Please see https://github.com/cilium/cilium/pull/15383#discussion_r602342047.

tklauser reviewed (approved):
This PR was a real pleasure to review - well structured and documented!

It would be nice to conserve the high-level overview from the PR description and some the commit messages to somewhere more visible. Maybe extract this into a small doc in `Documentation/concepts/` (as a follow-up)?

Some non-blocking comments inline, but LGTM overall.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -544,6 +550,17 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx,
+			    info->sec_label != HOST_ID &&
+			    info->sec_label != REMOTE_NODE_ID)
+				ctx->mark |= MARK_MAGIC_ENCRYPT;
+#endif /* ENABLE_WIREGUARD */
```
I'm a bit surprised to see that the mark configuration is occurring in different places from ipsec. Is there something drastically different between the two encryption models that explains this?

brb replied:
> Is there something drastically different between the two encryption models that explains this?

Yes, they differ quite a lot. In IPSec we need to return a packet to the stack to encrypt/decrypt it, and route it again to the same device. While in Wireguard it's enough just to redirect the packet to the Wireguard tunnel device (`cilium_wg0`) to encrypt it (decryption happens automatically when the `cilium_wg0` receives an encrypted packet).

brb replied:
Coming back to your question about the different places, I _think_ the mark could be set a bit later than it's now. Unfortunately, it cannot be set in the same place where the IPSec handling happens, as then the BPF host routing would make the packet to bypass the encryption (I haven't validated whether the BPF HR works with Wireguard though).

joestringer replied:
Yeah I guess the way I was thinking about this is that regardless of whether we need to forward this through the xfrm stack subsystem or route to the destination, we are passing the packets up to the stack to be redirected to the destination device and from that perspective, the main thing we need to do is set the mark before we hit that code and hopefully that could be done at one consistent "encryption" stage of the BPF code rather than having different behaviour/ordering depending on the encryption implementation. (This may not be 100% accurate particularly if we start to get into BPF host routing..).

The most obvious implication right now is just the L7 handling which I already mentioned below, but in general if the encryption is handled at different points of this program, then the other logic that surrounds the encrypt logic could either run or not run depending on the mode, and that could be a source of bugs. The fewer differences there are between the BPF implementations of the two encryption approaches, the better IMO.

brb replied:
Makes sense. For now I've moved setting the mark to the same place as we do for IPSec. We might need to move it higher once we start supporting BPF HR.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -165,6 +165,12 @@ static __always_inline int ipv6_l3_from_lxc(struct __ctx_buff *ctx,
+			if (info->tunnel_endpoint != 0 &&
+			    info->sec_label != HOST_ID &&
+			    info->sec_label != REMOTE_NODE_ID)
+				ctx->mark |= MARK_MAGIC_ENCRYPT;
```
Performing this before policy means that if we apply L7 egress policy, we're potentially setting the mark with all of the bits of both proxy redirect and encryption redirect.

joestringer replied:
^ Maybe short answer is "no L7 with wireguard" and we can decide whether to be strict about this in CLI parsing or just write it up as a limitation as part of wireguard beta with undefined behaviour.

brb replied:
My initial idea for setting the encrypt mark and passing to the stack instead of doing a redirect to the `cilium_wg0` was that (maybe) we could be compliant with L7. However, after taking a closer look I think I'm lacking an understanding of a packet life in our L7 world. For now, I'm going to add a check to make sure that Wireguard and L7 are mutually exclusive, and add the L7 support to the follow-up list.

brb replied:
Another high-level thought about supporting the L7 proxy. Maybe we could make https://github.com/cilium/cilium/pull/11279 work without `-m socket`, so that the mark usage is eliminated from the proxy path. If the proxy could retain the mark set by bpf_lxc and would pass the packet to the stack after it has validated it against L7 policies, we could make Wireguard easily to work with the proxy. Both, BPF TPROXY and Wireguard with KPR require >= 5.7 kernel.

joestringer replied:
> Maybe we could make #11279 work without -m socket, so that the mark usage is eliminated from the proxy path

We still need the mark to force the routing subsystem to route the proxy traffic to the local stack. Otherwise the destination address can be anything and the stack will just route the traffic to the destination instead of passing it to the proxy.

> If the proxy could retain the mark set by bpf_lxc and would pass the packet to the stack after it has validated it against L7 policies

Today, all packets from the proxy are marked with either `MARK_MAGIC_PROXY_INGRESS` or `MARK_MAGIC_PROXY_EGRESS` plus the source identity of the packet (see `inherit_identity_from_host()`). If you're suggesting that the proxy gets involved in setting the mark to forward to the encryption stack somehow, I think we're gonna have to think very carefully about this.

brb replied:
> If you're suggesting that the proxy gets involved in setting the mark to forward to the encryption stack somehow

Yep, that was my thinking. OK, I think after this PR has been merged, let's have a discussion offline how we can make it work. I assume that IPSec has the same problem then?

joestringer commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -339,6 +339,10 @@ func initKubeProxyReplacementOptions() (strict bool) {
 // handleNativeDevices tries to detect bpf_host devices (if needed).
 func handleNativeDevices(strict bool) {
+	if option.Config.EnableWireguard {
+		option.Config.DirectRoutingDevice = wireguardTypes.IfaceName
```
If we happen to be overriding user-specified values here, we should make that visible in the logs.

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -1433,6 +1445,24 @@ func runDaemon() {
 	iptablesManager.Init()
 
+	var wgAgent *wireguard.Agent
+	if option.Config.EnableWireguard {
```
Somewhere (maybe not here), we should check that IPsec and Wireguard are mutually exclusive.

borkmann replied:
Also, do we require to have remote node identities (given the earlier use of `REMOTE_NODE_ID`)?

joestringer replied:
not required, if the identity is seen in the ipcache then the code handles it. We *might* be able to optimize away that check but we'd have to look closely at the details. Unless we're really itching to minimize instructions, I would avoid looking into such an optimization for now.

joestringer commented on pkg/defaults/defaults.go:
```diff
@@ -408,4 +408,10 @@ const (
 	EnableBPFBypassFIBLookup = true
+
+	// WireguardSubnetV4 is a default wireguard tunnel subnet
+	WireguardSubnetV4 = "172.16.42.0/24"
```
Is there something special about this range? I feel like I've hit issues recently because of two other applications both thinking this particular /24 is a great range to use for independent reasons. (Well I think ultimately the reasons come down to a combination of RFC1918 and Douglas Adams, but that's not *so* informative by itself..)

brb replied:
Hehe :smile: I think I took it from the archlinux Wireguard example. I could change the 42 number to something less common, but ideally we should implement a subnet auto-detection for Wireguard (will add it to the follow-up list).

joestringer reviewed (commented):
I almost exclusively looked at the BPF side of this.

borkmann commented on bpf/lib/nodeport.h:
```diff
@@ -1846,6 +1849,12 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
+		 * Wireguard tunnel.
+		 */
+		ctx->mark |= MARK_MAGIC_ENCRYPT;
+#endif /* ENABLE_WIREGUARD */
```
Could we move all of these into small inline helpers that better document the intent? Would ipsec use the same?

borkmann replied:
Do we need to make KPR mandatory for WG?

brb replied:
No, KPR is not mandatory. This case is only for encrypting a fwd NodePort BPF request to a remote endpoint.

brb replied:
> Would ipsec use the same?

I haven't thoroughly checked the mark usage in our IPSec code.

jrfastab replied:
Mark field is also open coded like above in ipsec. Helpers would be nice.

kkourt commented on test/k8sT/Services.go:
```diff
@@ -2500,13 +2500,16 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
-					"Tests with Wireguard (L2-less)", func() {
+					"Tests with Wireguard tunnel to outside (L2-less, kube-wireguarder)", func() {
+						// This tests a case when Wireguard tunnel is established with a client
+						// located outside cluster. For this, we use kube-wireguarder, as the
```
nit: outside the cluster

kkourt commented on test/k8sT/Services.go:
```diff
@@ -2500,13 +2500,16 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+					"Tests with Wireguard tunnel to outside (L2-less, kube-wireguarder)", func() {
+						// This tests a case when Wireguard tunnel is established with a client
+						// located outside cluster. For this, we use kube-wireguarder, as the
+						// native implementation cannot be used to establish tunnels with
```
nit: s/native/cilium/  or s/native/native cilium/ ?

kkourt commented on test/k8sT/Services.go:
```diff
@@ -2597,10 +2569,40 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+							})
+
+							url := getHTTPLink(k8s1IP, data.Spec.Ports[0].NodePort)
+							testCurlFromOutside(url, 10, false)
```
AFAICT, the tests just check connectivity and not that the traffic is actually encrypted.

I also see that there is a:
> [ ] More tests

On the PR description.
Do the more tests include checking that the traffic is actually encrypted?

brb replied:
This is a very good point. I have elaborated a bit more on the testing in the follow-up issue: https://github.com/cilium/cilium/issues/15462.

brb replied:
I've added a test case with tcpdump to see whether we don't leak unencrypted packets. Currently, only for the case KPR case (we still miss a test case for pod2pod).

kkourt commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -417,6 +417,12 @@ data:
 {{- end }}
 
+{{- if hasKey .Values "wireguard" }}
+{{- if .Values.wireguard.enabled }}
```
Maybe not in this PR, but I think we should have something like `--encryption=ipvs|wireguard` and also have this reflected in the helm charts.

jrfastab replied:
I agree this would be better.

brb replied:
Created https://github.com/cilium/cilium/issues/15483 to address it before v1.10.

kkourt reviewed (approved):
Thanks for a great PR!

Have some nits here and there, but nothing blocking from my side.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -371,8 +378,11 @@ static __always_inline int ipv6_l3_from_lxc(struct __ctx_buff *ctx,
 #endif
 	} else
-#endif
-#endif
+#elif defined(ENABLE_WIREGUARD)
+	if (dst_remote_ep)
+		set_encrypt_mark(ctx);
```
Seems like this also needs the } else to make sure that the mark being set here doesn't immediately get overwritten..?

Or even if ENABLE_WIREGUARD and ENABLE_IDENTITY_MARK are mutually exclusive, that's not clear when reading this code right here.

brb replied:
Previously, I had the check in the `pkg/datapath/linux/config/config.go` to disable `ENABLE_IDENTITY_MARK` if Wireguard is enabled. But decided for simplicity reasons to follow the IPSec way (to add `else` as you suggested). Anyway, I expect a majority of Wireguard users to run with KPR enabled due to the kernel >= 5.4 constraint . So, with the socket-lb, setting the identity mark is not necessary for for the hairpin svc traffic.

joestringer commented on bpf/lib/nodeport.h:
```diff
@@ -1846,6 +1849,12 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
+		/* The request which is to be forwarded needs to go over the
+		 * Wireguard tunnel.
+		 */
+		ctx->mark |= MARK_MAGIC_ENCRYPT;
```
`set_encrypt_mark()`?

brb replied:
Oops, missed this case.

joestringer reviewed (approved):
I took another look at `bpf/`, couple of minor nits. Other than that, assuming we have a testing plan I'm OK with this from a mark perspective.

nathanjsweet commented on daemon/cmd/daemon_main.go:
```diff
@@ -1433,6 +1445,35 @@ func runDaemon() {
+		wgAgent, err = wireguard.NewAgent(privateKeyPath,
+			option.Config.WireguardSubnetV4, option.Config.WireguardSubnetV6)
+		if err != nil {
+			log.WithError(err).Fatal("Failed to initialize wireguard")
```
Why isn't this fatal?

brb replied:
This is fatal: `log.WithError(err).>>>Fatal<<<("Failed to initialize wireguard")`.

nathanjsweet commented on daemon/cmd/daemon_main.go:
```diff
@@ -1484,6 +1525,11 @@ func runDaemon() {
 	restoreComplete := d.initRestore(restoredEndpoints)
+	if wgAgent != nil {
+		if err := wgAgent.RestoreFinished(); err != nil {
+			log.WithError(err).Error("Failed to set up wireguard peers")
```
Why isn't this fatal?

brb replied:
It's not the end of the world if the WG agent restore fails. The worst what can happen is that the WG device will have no longer existing peers.

nathanjsweet commented on operator/main.go:
```diff
@@ -403,10 +415,17 @@ func onOperatorStartLeading(ctx context.Context) {
+	if wgOperator != nil {
+		<-k8sCiliumNodesCacheSynced
+		if err := wgOperator.RestoreFinished(); err != nil {
+			log.WithError(err).Warn("Failed to allocate wireguard IP addrs")
```
Ditto.

brb replied:
Same reasoning as with the WG agent restore. Just that the worst case is that it will take a bit longer until a new node gets a Wireguard tunnel IP allocated.

nathanjsweet reviewed (commented):
A couple of questions.
